### PR TITLE
Added modelToIgnore option which…

### PR DESCRIPTION
**Pull request by aquid**
… contains a list of models that are to be prevented from being published in loopback angular SDK.

@bajtos You will have to update the package.json with right version number , once you accept and publish the PR,
https://github.com/strongloop/loopback-sdk-angular/pull/224

### CODE REVIEW

**Comment by slnode** [comment]
Can one of the admins verify this patch? To accept patch and trigger a build add comment ".*ok\W+to\W+test.*"

**Comment by aquid** [comment]
@bajtos Any updates on this?

**Comment by slnode** [comment]
Can one of the admins verify this patch?

File: package.json [hunk 1]
```diff
@@ -23,7 +23,7 @@
     "test": "grunt test"
   },
   "dependencies": {
-    "loopback-sdk-angular": "^1.2.0"
+    "loopback-sdk-angular": "https://github.com/aquid/loopback-sdk-angular/tarball/master"
```
Comment (bajtos) [inline package.json]: please revert

Comment (bajtos) [inline package.json]: Even better, change the dependency version spec to `^1.10.0` to ensure we have a version that supports `modelsToIgnore` option.

File: tasks/loopback_angular.js [hunk 2]
```diff
@@ -32,7 +32,8 @@ module.exports = function(grunt) {
     // Merge task-specific and/or target-specific options with these defaults.
     var options = this.options({
       ngModuleName: 'lbServices',
-      apiUrl: undefined
+      apiUrl: undefined,
+      modelsToIgnore: []
```
Comment (bajtos) [inline tasks/loopback_angular.js]: I believe `options.modelsToIgnore` should be handled by the generator, so let's use that as the default value and revert this change.

File: tasks/loopback_angular.js [hunk 3]
```diff
@@ -45,6 +46,9 @@ module.exports = function(grunt) {
     if (!options.output)
       grunt.fail.warn('Missing mandatory option "output".');
 
+    if(!options.modelsToIgnore)
+      grunt.log.writeln('No models to ingnore, Generating endpoints for each models');
```
Comment (bajtos) [inline tasks/loopback_angular.js]: I prefer the tools to be quite and print to console only when there is something needing my attention.

Please revert this change too.

**Comment by bajtos** [comment]
@aquid please add a unit-test to verify this change. IIRC, you will need to add a new section to https://github.com/strongloop/grunt-loopback-sdk-angular/blob/8ab1baf61d12962c83394e24af8942d321c1123a/Gruntfile.js#L37-L54 and then a test case here: https://github.com/strongloop/grunt-loopback-sdk-angular/blob/8ab1baf61d12962c83394e24af8942d321c1123a/test/loopback_sdk_angular_test.js#L39-L54

**Comment by bajtos** [comment]
@aquid also please rebase your patch on top of the current `master` and make sure the commit messages are following the [50/72 rule](http://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html).

**Comment by bajtos** [comment]
@aquid ping, what's the status of this pull request? Are you still keen to get it finished?

**Comment by aquid** [comment]
@bajtos Apologies for the delay but, I am on it. Just need some time to add unit tests. Have to finish on some other projects, once that are done will work on the PR.

**Comment by aquid** [comment]
@bajtos Is there a way to test `modelsToIgnore` in our unit test? Right now I can just test the equality of `moduleName` and `baseUrl` apart from this I don't think we can test anything else.

What I have done till now is this

```
Gruntfile.js: 

ignore_models_option: {
        options: {
          output: 'tmp/ignore_models_option',
          ngModuleName: 'modelsToIgnoreService',
          modelsToIgnore : ['customModel_1','customModel_2']
        }
      }


loopback_sdk_angular_test.js:

ignore_models_option: function(test) {
    var script = grunt.file.read('tmp/ignore_models_option');
    test.equal(parse.moduleName(script), 'modelsToIgnoreService');
    // This won't work as there's no export for this.
    // test.equal(parse.modelsToIgnore(script), ['customModel_1','customModel_2']);
    test.done();
  }

```

Will that be enough for the tests?

**Comment by aquid** [comment]
@bajtos I have updated the PR. Please let me know if there's anything else that I should add here.

**Comment by bajtos** [comment]
> Will that be enough for the tests?
> ```
    // This won't work as there's no export for this.
    // test.equal(parse.modelsToIgnore(script), ['customModel_1','customModel_2']);
```

You can check that the script does not contain any of these models. Something along the following lines:

```js
// assuming User model is ignored
test.ok(script.indexOf('User') === -1);
```

Also make sure to see your test fail. For example, the currently proposed test excluding `customModel_1` will always pass, because there is no `customModel_1` in the fixture app.

File: package.json [hunk 4]
```diff
@@ -32,8 +32,8 @@
     "grunt-contrib-clean": "^1.0.0",
     "grunt-contrib-jshint": "^1.0.0",
     "grunt-contrib-nodeunit": "^1.0.0",
-    "loopback": "^1.7.4",
-    "loopback-datasource-juggler": "^1.3.10"
+    "loopback": "^2.0.0",
```
Comment (aquid) [inline package.json]: @bajtos Changed the version because of this [issue](https://github.com/strongloop/loopback-sdk-angular/issues/236)

**Comment by aquid** [comment]
@bajtos Updated PR with custom models initialization and tests to verify if models are ignored properly.

File: test/fixtures/app.js [hunk 5]
```diff
@@ -13,6 +13,32 @@ app.dataSource('mail', { connector: 'mail', defaultForType: 'mail' });
 // Attach all built-in models
 loopback.autoAttach();
 
+// Get loopback model reference 
+var Model = loopback.Model;
+
+// Create custom models extending Model 
+var CustomModel_1 = Model.extend('CustomModel_1', {}, {
+  trackChanges: true,
```
Comment (bajtos) [inline test/fixtures/app.js]: Why is `trackChanges` needed? AFAICT, your test is not using this feature at all.

File: test/fixtures/app.js [hunk 6]
```diff
@@ -13,6 +13,32 @@ app.dataSource('mail', { connector: 'mail', defaultForType: 'mail' });
...
+
+var CustomModel_2 = Model.extend('CustomModel_2', {}, {
+  trackChanges: true,
+});
+
+var CustomModel_3 = Model.extend('CustomModel_3', {}, {
+  trackChanges: true,
+});
+
+// Attach custom model to memory(db)
```
Comment (bajtos) [inline test/fixtures/app.js]: Please replace `.attachTo` and `app.model` with the following:

```js
app.model(CustomModel, { dataSource: 'db' });
```

File: test/loopback_sdk_angular_test.js [hunk 7]
```diff
@@ -51,5 +51,14 @@ exports.loopback_angular = {
     test.equal(parse.moduleName(script), 'customServices');
     test.equal(parse.baseUrl(script), 'http://custom/api');
     test.done();
+  },
+  ignore_models_option: function(test) {
```
Comment (bajtos) [inline test/loopback_sdk_angular_test.js]: Add an empty line between test cases to visually delimit them. See e.g. L47-L48.

File: test/loopback_sdk_angular_test.js [hunk 8]
```diff
@@ -51,5 +51,14 @@ exports.loopback_angular = {
...
+    var script = grunt.file.read('tmp/ignore_models_option');
+    test.equal(parse.moduleName(script), 'modelsToIgnoreService');
+    test.equal(parse.baseUrl(script), 'http://ignore_models/api');
```
Comment (bajtos) [inline test/loopback_sdk_angular_test.js]: These two checks (module name, base url) are not relevant for this test and are already covered by `custom_options` test, please remove.

**Comment by bajtos** [comment]
@aquid almost there :) See my comment above, and also please edit the first line of the commit message to mention "modelsToIgnore option".

**Comment by aquid** [comment]
@bajtos Hopefully fixed everything this time. Waiting for your response. (Fingers crossed).

**Comment by bajtos** [comment]
@slnode ok to test

**Comment by slnode** [comment]
Can one of the admins verify this patch? To accept patch and trigger a build add comment ".*ok\W+to\W+test.*"

**Comment by slnode** [comment]
Can one of the admins verify this patch?

**Comment by bajtos** [comment]
Landed, thank you for the contribution!

**Comment by bajtos** [comment]
Released as `grunt-loopback-sdk-angular@1.3.0`.
